File: kotlin/lesson 10/FibonacciNumberRecursive.kt
// https://leetcode.com/problems/fibonacci-number/

class Solution {
    fun fib(n: Int): Int {
        if (n <= 1) return n        
        return fib(n, IntArray(n + 1))
    }
    
    fun fib(n: Int, dp: IntArray): Int {
        if (n <= 1) return n
        if (dp[n] != 0) return dp[n]
        dp[n] = fib(n - 1, dp) + fib(n - 2, dp)
        return dp[n]
    }
}

File: python/lesson 10/fibonacci_number_recursive.py
# https://leetcode.com/problems/fibonacci-number/

class Solution:
    def fib(self, n: int) -> int:
        if n <= 1: return n
        dp = [None] * (n + 1)
        dp[0] = 0
        dp[1] = 1
        return self.fibImpl(n, dp)
    
    def fibImpl(self, n: int, dp: List[int]):
        if (dp[n] != None): return dp[n]
        dp[n] = self.fibImpl(n - 1, dp) + self.fibImpl(n - 2, dp)
        return dp[n]

